app/src/main/java/mozilla/lockbox/presenter/RoutePresenter.kt

```diff
@@ -55,33 +58,51 @@ class RoutePresenter(
     override fun onViewReady() {
         navController = Navigation.findNavController(activity, R.id.fragment_nav_host)
 
-        // Mediates credentials from the AccountStore, into the DataStore.
-        Observables.combineLatest(
-                accountStore.syncCredentials.filterNotNull(),
-                dataStore.state)
-            .subscribe {
-                val action = when (it.second) {
-                    is State.Unprepared -> DataStoreAction.UpdateCredentials(it.first)
-                    is State.Unlocked -> DataStoreAction.Sync
-                    else -> return@subscribe
-                }
-                dispatcher.dispatch(action)
-            }
-            .addTo(compositeDisposable)
+        // Make the routing contingent on the state of the dataStore.
+        val dataStoreRoutes = dataStore.state
+            .map(this::dataStoreToRouteActions)
+            .filterNotNull()
 
-        val useTestData = lifecycleStore.lifecycleFilter
-            .filter { it == LifecycleAction.UseTestData }
-            .doOnNext {
-                dataStore.resetSupport(FixedDataStoreSupport())
-            }
-            .map { RouteAction.ItemList }
+        // Moves credentials from the AccountStore, into the DataStore.
+        Observables.combineLatest(accountStore.syncCredentials, dataStore.state)
+            .map(this::accountToDataStoreActions)
+            .filterNotNull()
+            .subscribe(dispatcher::dispatch)
+            .addTo(compositeDisposable)
 
         routeStore.routes
-            .mergeWith(useTestData)
+            .mergeWith(dataStoreRoutes)
+            .observeOn(mainThread())
             .subscribe(this::route)
             .addTo(compositeDisposable)
     }
 
+    private fun dataStoreToRouteActions(storageState: State): Optional<RouteAction> {
+        return when (storageState) {
+            is State.Unlocking -> null
+            is State.Unlocked -> RouteAction.ItemList
+            is State.Locked -> RouteAction.LockScreen
+            is State.Unprepared -> RouteAction.Welcome
+            else -> RouteAction.LockScreen
+        }.asOptional()
+    }
+
+    private fun accountToDataStoreActions(it: Pair<Optional<SyncCredentials>, State>): Optional<DataStoreAction> {
+        val (opt, state) = it
+
+        val credentials = opt.value ?: return DataStoreAction.Reset.asOptional()
+
+        if (state != State.Unprepared) {
+            return Optional(null)
+        }
+
+        if (credentials is FixedSyncCredentials) {
+            dataStore.resetSupport(FixedDataStoreSupport.shared)
```

Comment:
why are we calling this directly rather than doing it via an action? is it too heavyweight to pass around a `support` reference in an action?

Reply:
Fixed.